// @terminal-packages/sdk/dist/providers/http/terminal-http-provider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_providers_1 = require("web3-providers");
const utils_1 = require("../../shared/utils");
const terminal_base_provider_1 = require("../base/terminal-base-provider");
const enums_1 = require("../enums");
class TerminalHttpProvider extends terminal_base_provider_1.TerminalBaseProvider {
    constructor(options) {
        super(options);
        // @ts-ignore
        if (options.customHttpProvider) {
            if (options.source !== enums_1.SourceType.MetaMask) {
                const customProviderOptions = options;
                this.host = customProviderOptions.customHttpProvider.host;
                if (!this.host || this.host.length === 0) {
                    throw new Error('Your custom http provider host is not defined or is set as a empty string.');
                }
                this._provider = customProviderOptions.customHttpProvider;
            }
            else {
                this.buldMetaMaskHttpProvider();
            }
        }
        else {
            const basicProviderOptions = options;
            this.host = basicProviderOptions.host;
            if (!this.host || this.host.length === 0) {
                throw new Error('The host is not defined or is set as a empty string');
            }
            this._provider = new web3_providers_1.HttpProvider(this.host, basicProviderOptions.options);
        }
        this.setChainId(this._provider);
        this.connected = true;
    }
    /**
     * Gets the provider and makes sure it comes back as `WebsocketProvider` type
     */
    get httpProvider() {
        if (!this._provider) {
            throw new Error('Provider is not set');
        }
        return this._provider;
    }
    /**
     * Supports subsciptions
     */
    supportsSubscriptions() {
        return this.httpProvider.supportsSubscriptions();
    }
    /**
     * Disconnect (wont be a thing here but its based off the interface)
     */
    disconnect() {
        return this.httpProvider.disconnect();
    }
    /**
     * Build MetaMask http provider
     */
    buldMetaMaskHttpProvider() {
        const _window = utils_1.Utils.getWindow();
        if (!_window) {
            throw new Error('MetaMask provider can only be used on a webapp');
        }
        if (!_window.ethereum) {
            throw new Error('MetaMask is not installed so can not log MetaMask jsonrpc calls');
        }
        // we set the host here but `window.ethereum` may not of been
        // loaded yet so when we go get the host we double check
        this.host = _window.ethereum.networkVersion;
        // @ts-ignore
        this._provider = {
            send: (method, parameters) => __awaiter(this, void 0, void 0, function* () {
                return yield _window.ethereum.send(method, parameters);
            }),
            disconnect: () => false,
            supportsSubscriptions: () => false
        };
    }
}
exports.TerminalHttpProvider = TerminalHttpProvider;
